text/0000-safer-transmute.md

```diff
@@ -0,0 +1,1796 @@
+```rust
+pub evil_constructor<T>(src: T) -> NonEmptySlice<'static, u8>
+where
+    T: TransmuteInto<NonEmptySlice<'static, u8>>,
+{
+    src.transmute_into()
+}
+
+evil_constructor(0u128).first() // muahaha!
+```
+This safety hazard is not materially different from the one that would be induced if the type author implemented `DerefMut<Target=NonEmptySliceInner>` for `NonEmptySlice`, or made the `private` module `pub`, or otherwise explicitly provided outsiders with unrestricted mutable access to `data`.
+
+##### Recommendation
+We recommend that that implementers of this RFC initially simplify constructability by:
+ - adopting our simplified definition of constructability
+ - demoting `NeglectStability` to unsafe status (i.e., not implementing `SafeTransmuteOptions` for `NeglectStability`; *only* `TransmuteOptions`)
+ - advise users that implementing the stability declaration traits on types that are not fully-implicitly constructable will be a compiler-error will be a compiler error (i.e., these traits must not be implemented on types exploiting the pub-in-priv trick)
+
+If and when the implementation of `TransmuteFrom` encodes our complete definition of constructability, `NeglectStability` shall become a safe transmute option.
+
+### Minimal Useful Stabilization Surface
+
+Stabilizing *only* these three items of the Initial Smart Implementation will cover most use-cases:
+  - `TransmuteFrom`
+  - `#[derive(PromiseTransmutableFrom)]`
+  - `#[derive(PromiseTransmutableInto)]`
+
+(If the [`PromiseTransmutable` shorthand extension][extension-promisetransmutable-shorthand] is accepted, this may be further reduced to just *two* items: `TransmuteFrom` and `#[derive(PromiseTransmutable)]`.)
+
+Stabilizing `TransmuteOptions` and `SafeTransmuteOptions` will additionally allow end-users to build generic abstractions over `TransmuteFrom` (e.g., slice casting abstractions).
+
+Stabilizing `PromiseTransmutableFrom` and `PromiseTransmutableInto` will additionally allow end-users make limited stability promises, and to make stability promises for types where `derive` is too restrictive (namely, types containing `PhantomData`).
+
+Stabilizing `TransmuteInto` and `NeglectStability` will additionally allow end-users to implement `PromiseTransmutableFrom` in cases where the `Archetype`'s trait bounds must be repeated for lack of [implied bounds](https://github.com/rust-lang/rust/issues/44491); e.g.:
+```rust
+impl<T, N> PromiseTransmutableFrom for GenericArray<T, N>
+where
+    T: PromiseTransmutableFrom,
+    N: ArrayLength<T>,
+
+    // for lack of implied bounds, we must repeat the bounds on `Archetype`:
+    GenericArray<T::Archetype, N>
+        : TransmuteInto<Self, NeglectStability>
+        + PromiseTransmutableFrom,
+{
+    type Archetype = GenericArray<T::Archetype, N>;
+}
```

Comment:
Is it feasible to limit uses of `NeglectStability` to *only* implementations of `PromiseTransmutable*`? (Similar to how `!` is currently stably usable in one context, but not others?)